error building open3D with CUDA_MODULE compatibility

**Describe the bug**
A clear and concise description of what the bug is.

**To Reproduce**
Steps to reproduce the behavior:
1. cd open3d
2. mkdir build
3. cd build
4. cmake -DCMAKE_INSTALL_PREFIX=installpath-open3d -DBUILD_CUDA_MODULE=TRUE ..
5. make -j8
6. See error

**Expected behavior**
I get the Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) while compiling Kernel/ReductionCUDA 

**Screenshots**
![Screenshot from 2020-05-06 13-20-17](https://user-images.githubusercontent.com/3373611/81171374-92df4f80-8f9c-11ea-9904-af1773457f2c.png)


**Environment (please complete the following information):**
 - OS: Ubuntu 16.04
 - Python version: Python 3.6 (conda)
 - Open3D version: 0.9.0
 - Is this remote workstation?: no
 - How did you install Open3D?: build_from_source (GCC-5 and also try GCC-7)

**Additional context**
I'm using CUDA toolkit 10.2 using a CUDA device with arch 5 compatibility

I'm actually experiencing this issue right now trying to build from source on Ubuntu 18.04. My GPU is an NVIDIA GeForce GTX 1080. I am running CUDA version 10.0.326. Posting the entirety of my output:

`[  0%] Built target tinyobjloader
[  2%] Built target ext_turbojpeg
[  3%] Built target glew
[  4%] Built target jsoncpp
[  8%] Built target qhullstatic_r
[ 12%] Built target glfw
[ 18%] Built target qhullcpp
[ 19%] Built target tinyfiledialogs
[ 24%] Built target zlib
Custom target build_all_3rd_party_libs reached.
[ 24%] Built target EncodeShader
[ 24%] Built target Python
[ 24%] Built target build_all_3rd_party_libs
[ 29%] Built target benchmark
[ 33%] Built target png
[ 33%] Built target ShaderFileTarget
[ 34%] Built target Integration
Scanning dependencies of target Core
[ 34%] Built target Camera
[ 34%] Built target Odometry
[ 35%] Built target ColorMap
[ 35%] Built target benchmark_main
[ 37%] Built target Registration
[ 39%] Built target Utility
[ 40%] Building CXX object src/Open3D/Geometry/CMakeFiles/Geometry.dir/TriangleMesh.cpp.o
[ 41%] Building CXX object src/Open3D/IO/CMakeFiles/IO.dir/FileFormat/FilePNG.cpp.o
[ 41%] Building CXX object src/Open3D/IO/CMakeFiles/IO.dir/FileFormat/FilePTS.cpp.o
[ 41%] Building CXX object src/Open3D/Geometry/CMakeFiles/Geometry.dir/SurfaceReconstructionPoisson.cpp.o
[ 41%] Building CXX object src/Open3D/IO/CMakeFiles/IO.dir/FileFormat/FileSTL.cpp.o
[ 41%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Shader/PickingShader.cpp.o
[ 42%] Building CXX object src/Open3D/Core/CMakeFiles/Core.dir/AdvancedIndexing.cpp.o
[ 42%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Shader/RGBDImageShader.cpp.o
[ 42%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Shader/ShaderWrapper.cpp.o
[ 43%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Shader/Simple2DShader.cpp.o
[ 43%] Building CXX object src/Open3D/IO/CMakeFiles/IO.dir/FileFormat/FileTUM.cpp.o
[ 43%] Building CXX object src/Open3D/Core/CMakeFiles/Core.dir/Indexer.cpp.o
[ 43%] Building CXX object src/Open3D/Geometry/CMakeFiles/Geometry.dir/TriangleMeshDeformation.cpp.o
[ 43%] Building CXX object src/Open3D/Geometry/CMakeFiles/Geometry.dir/TriangleMeshFactory.cpp.o
[ 43%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Shader/SimpleBlackShader.cpp.o
[ 44%] Building CXX object src/Open3D/IO/CMakeFiles/IO.dir/FileFormat/FileXYZ.cpp.o
[ 44%] Building CXX object src/Open3D/Core/CMakeFiles/Core.dir/Tensor.cpp.o
[ 44%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Shader/SimpleShader.cpp.o
[ 44%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Shader/TexturePhongShader.cpp.o
[ 45%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Shader/TextureSimpleShader.cpp.o
[ 45%] Building CXX object src/Open3D/IO/CMakeFiles/IO.dir/FileFormat/FileXYZN.cpp.o
[ 45%] Building CXX object src/Open3D/IO/CMakeFiles/IO.dir/FileFormat/FileXYZRGB.cpp.o
[ 45%] Building CXX object src/Open3D/Core/CMakeFiles/Core.dir/Kernel/IndexGetSetCPU.cpp.o
[ 45%] Building CXX object src/Open3D/Geometry/CMakeFiles/Geometry.dir/TriangleMeshSimplification.cpp.o
[ 46%] Building CXX object src/Open3D/Geometry/CMakeFiles/Geometry.dir/TriangleMeshSubdivide.cpp.o
[ 46%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Utility/ColorMap.cpp.o
[ 47%] Building C object src/Open3D/IO/CMakeFiles/IO.dir/__/__/__/3rdparty/rply/rply/rply.c.o
[ 47%] Building CXX object src/Open3D/Core/CMakeFiles/Core.dir/Kernel/UnaryEWCPU.cpp.o
[ 47%] Building C object src/Open3D/IO/CMakeFiles/IO.dir/__/__/__/3rdparty/liblzf/liblzf/lzf_c.c.o
[ 47%] Building C object src/Open3D/IO/CMakeFiles/IO.dir/__/__/__/3rdparty/liblzf/liblzf/lzf_d.c.o
[ 52%] Built target IO
[ 53%] Building CXX object src/Open3D/Core/CMakeFiles/Core.dir/Kernel/BinaryEWCPU.cpp.o
[ 53%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Utility/DrawGeometry.cpp.o
[ 54%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Utility/GLHelper.cpp.o
[ 54%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Utility/PointCloudPicker.cpp.o
[ 54%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Utility/SelectionPolygon.cpp.o
[ 54%] Building CXX object src/Open3D/Core/CMakeFiles/Core.dir/Kernel/ReductionCPU.cpp.o
[ 54%] Building CUDA object src/Open3D/Core/CMakeFiles/Core.dir/Kernel/IndexGetSetCUDA.cu.o
[ 54%] Building CXX object src/Open3D/Geometry/CMakeFiles/Geometry.dir/VoxelGrid.cpp.o
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Device.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Blob.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSetCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
[ 54%] Building CXX object src/Open3D/Geometry/CMakeFiles/Geometry.dir/VoxelGridFactory.cpp.o
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexGetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexGetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE0_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexGetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE1_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexGetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE2_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexGetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE3_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexSetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexSetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE0_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexSetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE1_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexSetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE2_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_
/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/CUDALauncher.cuh(53): Error: Formal parameter space overflowed (4128 bytes required, max 4096 bytes allowed) in function _ZN6open3d6kernel17ElementWiseKernelILl128ELl4EZNS0_12CUDALauncher27LaunchAdvancedIndexerKernelIZZZNS0_12IndexSetCUDAERKNS_6TensorERS4_RKSt6vectorIS4_SaIS4_EERKNS_10SizeVectorESF_ENKUlvE_clEvENKUlvE3_clEvEUlPKvPvE_EEvRKNS_15AdvancedIndexerET_EUllE_EEvlT1_

src/Open3D/Core/CMakeFiles/Core.dir/build.make:309: recipe for target 'src/Open3D/Core/CMakeFiles/Core.dir/Kernel/IndexGetSetCUDA.cu.o' failed
make[2]: *** [src/Open3D/Core/CMakeFiles/Core.dir/Kernel/IndexGetSetCUDA.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 54%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Utility/SelectionPolygonVolume.cpp.o
[ 55%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/RenderOption.cpp.o
[ 55%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/RenderOptionWithEditing.cpp.o
CMakeFiles/Makefile2:1298: recipe for target 'src/Open3D/Core/CMakeFiles/Core.dir/all' failed
make[1]: *** [src/Open3D/Core/CMakeFiles/Core.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 55%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/ViewControl.cpp.o
[ 55%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/ViewControlWithCustomAnimation.cpp.o
[ 56%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/ViewControlWithEditing.cpp.o
[ 56%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/ViewParameters.cpp.o
[ 56%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/ViewTrajectory.cpp.o
[ 56%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/Visualizer.cpp.o
[ 57%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/VisualizerCallback.cpp.o
[ 57%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/VisualizerRender.cpp.o
[ 57%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/VisualizerWithCustomAnimation.cpp.o
[ 63%] Built target Geometry
[ 63%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/VisualizerWithEditing.cpp.o
[ 64%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/VisualizerWithKeyCallback.cpp.o
[ 64%] Building CXX object src/Open3D/Visualization/CMakeFiles/Visualization.dir/Visualizer/VisualizerWithVertexSelection.cpp.o
[ 66%] Built target Visualization
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
`

@yxlao When I try the modification you suggested I get a whole bunch of other errors, though most occur when it is failing after the initial error.

```
[ 65%] Built target Visualization
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/SizeVector.h:34,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/Reduction.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/ReductionCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Device.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Blob.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEW.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Device.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Blob.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSetCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Device.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Blob.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Device.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Blob.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSetCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(811): error: explicit specialization of function "open3d::Tensor::Tensor(const std::vector<T, std::allocator<T>> &, const open3d::SizeVector &, open3d::Dtype, const open3d::Device &) [with T=bool]" must precede its first use (
(266): here)

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(815): error: expected a ";"

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(837): warning: parsing restarts here after previous syntax error

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(841): error: "Tensor" is not a class or function template name in the current scope

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(841): error: expected a ";"

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(59): warning: parsing restarts here after previous syntax error

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(61): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(62): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(63): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(64): error: identifier "SizeVector" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(65): error: identifier "SizeVector" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(68): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(69): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(70): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(71): error: identifier "SizeVector" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(72): error: identifier "SizeVector" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/IndexGetSet.h(75): error: expected a declaration

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(59): warning: parsing restarts here after previous syntax error

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(430): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(440): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(10118): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(10188): error: identifier "cuuint32_t" is undefined

19 errors detected in the compilation of "/tmp/tmpxft_00007cd0_00000000-4_IndexGetSetCUDA.cpp4.ii".
src/Open3D/Core/CMakeFiles/Core.dir/build.make:309: recipe for target 'src/Open3D/Core/CMakeFiles/Core.dir/Kernel/IndexGetSetCUDA.cu.o' failed
make[2]: *** [src/Open3D/Core/CMakeFiles/Core.dir/Kernel/IndexGetSetCUDA.cu.o] Error 2
src/Open3D/Core/CMakeFiles/Core.dir/build.make:309: recipe for target 'src/Open3D/Core/CMakeFiles/Core.dir/Kernel/IndexGetSetCUDA.cu.o' failed
make[2]: *** [src/Open3D/Core/CMakeFiles/Core.dir/Kernel/IndexGetSetCUDA.cu.o] Error 2
make[2]: *** Waiting for unfinished jobs....
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/SizeVector.h:34,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/Reduction.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/ReductionCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(811): error: explicit specialization of function "open3d::Tensor::Tensor(const std::vector<T, std::allocator<T>> &, const open3d::SizeVector &, open3d::Dtype, const open3d::Device &) [with T=bool]" must precede its first use (
(266): here)

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(815): error: expected a ";"

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(837): warning: parsing restarts here after previous syntax error

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(841): error: "Tensor" is not a class or function template name in the current scope

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(841): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(93): error: hash is not a template

/usr/include/c++/7/bits/unordered_set.h(98): error: __uset_hashtable is not a template

/usr/include/c++/7/bits/unordered_set.h(105): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(105): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(105): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(106): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(106): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(106): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(107): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(107): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(107): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(108): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(108): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(108): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(109): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(109): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(109): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(114): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(114): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(114): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(115): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(115): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(115): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(116): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(116): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(116): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(117): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(117): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(117): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(118): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(118): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(118): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(119): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(119): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(119): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(120): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(120): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(120): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(121): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(121): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(121): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(122): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(122): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(122): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(123): error: a class or namespace qualified name is required

/usr/include/c++/7/bits/unordered_set.h(123): error: global-scope qualifier (leading "::") is not allowed

/usr/include/c++/7/bits/unordered_set.h(123): error: expected a ";"

/usr/include/c++/7/bits/unordered_set.h(144): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(145): error: identifier "hasher" is undefined

/usr/include/c++/7/bits/unordered_set.h(146): error: identifier "key_equal" is undefined

/usr/include/c++/7/bits/unordered_set.h(147): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(166): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(167): error: identifier "hasher" is undefined

/usr/include/c++/7/bits/unordered_set.h(168): error: identifier "key_equal" is undefined

/usr/include/c++/7/bits/unordered_set.h(169): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(184): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(194): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(204): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(219): error: initializer_list is not a template

/usr/include/c++/7/bits/unordered_set.h(219): error: identifier "value_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(220): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(221): error: identifier "hasher" is undefined

/usr/include/c++/7/bits/unordered_set.h(222): error: identifier "key_equal" is undefined

/usr/include/c++/7/bits/unordered_set.h(223): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(227): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(227): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(231): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(231): error: identifier "hasher" is undefined

/usr/include/c++/7/bits/unordered_set.h(232): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(238): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(239): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(245): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(245): error: identifier "hasher" is undefined

/usr/include/c++/7/bits/unordered_set.h(246): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(250): error: initializer_list is not a template

/usr/include/c++/7/bits/unordered_set.h(251): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(252): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(256): error: initializer_list is not a template

/usr/include/c++/7/bits/unordered_set.h(257): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(257): error: identifier "hasher" is undefined

/usr/include/c++/7/bits/unordered_set.h(258): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(282): error: initializer_list is not a template

/usr/include/c++/7/bits/unordered_set.h(289): error: identifier "allocator_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(301): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(306): error: identifier "size_type" is undefined

/usr/include/c++/7/bits/unordered_set.h(317): error: identifier "iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(321): error: identifier "const_iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(331): error: identifier "iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(335): error: identifier "const_iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(344): error: identifier "const_iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(352): error: identifier "const_iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(374): error: identifier "iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(400): error: identifier "iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(401): error: identifier "const_iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(418): error: identifier "iterator" is undefined

/usr/include/c++/7/bits/unordered_set.h(419): error: "value_type" is not a type name

Error limit reached.
100 errors detected in the compilation of "/tmp/tmpxft_00007ce2_00000000-4_ReductionCUDA.cpp4.ii".
Compilation terminated.
src/Open3D/Core/CMakeFiles/Core.dir/build.make:348: recipe for target 'src/Open3D/Core/CMakeFiles/Core.dir/Kernel/ReductionCUDA.cu.o' failed
make[2]: *** [src/Open3D/Core/CMakeFiles/Core.dir/Kernel/ReductionCUDA.cu.o] Error 4
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Device.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Blob.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEW.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(811): error: explicit specialization of function "open3d::Tensor::Tensor(const std::vector<T, std::allocator<T>> &, const open3d::SizeVector &, open3d::Dtype, const open3d::Device &) [with T=bool]" must precede its first use (
(266): here)

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(815): error: expected a ";"

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(837): warning: parsing restarts here after previous syntax error

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(841): error: "Tensor" is not a class or function template name in the current scope

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(841): error: expected a ";"

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(59): warning: parsing restarts here after previous syntax error

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(430): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(440): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(10118): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(10188): error: identifier "cuuint32_t" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(153): error: identifier "UnaryEWOpCode" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(167): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/UnaryEWCUDA.cu(180): error: name followed by "::" must be a class or namespace name

40 errors detected in the compilation of "/tmp/tmpxft_00007cd8_00000000-4_UnaryEWCUDA.cpp4.ii".
src/Open3D/Core/CMakeFiles/Core.dir/build.make:322: recipe for target 'src/Open3D/Core/CMakeFiles/Core.dir/Kernel/UnaryEWCUDA.cu.o' failed
make[2]: *** [src/Open3D/Core/CMakeFiles/Core.dir/Kernel/UnaryEWCUDA.cu.o] Error 2
In file included from /home/wnitsch/src/Open3D/3rdparty/Eigen/Eigen/Core:257:0,
                 from /home/wnitsch/src/Open3D/src/Open3D/Utility/Console.h:29,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Device.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Blob.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h:33,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h:31,
                 from /home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu:27:
/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/host_defines.h:54:2: warning: #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead." [-Wcpp]
 #warning "host_defines.h is an internal header file and must not be used directly.  This file will be removed in a future CUDA release.  Please use cuda_runtime_api.h or cuda_runtime.h instead."
  ^~~~~~~
/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(811): error: explicit specialization of function "open3d::Tensor::Tensor(const std::vector<T, std::allocator<T>> &, const open3d::SizeVector &, open3d::Dtype, const open3d::Device &) [with T=bool]" must precede its first use (
(266): here)

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(815): error: expected a ";"

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(837): warning: parsing restarts here after previous syntax error

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(841): error: "Tensor" is not a class or function template name in the current scope

/home/wnitsch/src/Open3D/src/Open3D/Core/Tensor.h(841): error: expected a ";"

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(60): warning: parsing restarts here after previous syntax error

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(62): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(63): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(64): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(65): error: identifier "BinaryEWOpCode" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(68): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(69): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(70): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(71): error: identifier "BinaryEWOpCode" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(74): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(74): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(74): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(75): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(75): error: identifier "BinaryEW" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(78): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(78): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(78): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(79): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(79): error: identifier "BinaryEW" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(82): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(82): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(82): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(83): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(83): error: identifier "BinaryEW" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(86): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(86): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(86): error: identifier "Tensor" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(87): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(87): error: identifier "BinaryEW" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEW.h(90): error: expected a declaration

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(59): warning: parsing restarts here after previous syntax error

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(430): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(440): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(10118): error: identifier "cuuint32_t" is undefined

/usr/local/cuda-10.0/bin/../targets/x86_64-linux/include/cuda.h(10188): error: identifier "cuuint32_t" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(149): error: identifier "BinaryEWOpCode" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(152): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(160): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(167): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(175): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(182): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(189): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(196): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(203): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(210): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(225): error: identifier "BinaryEWOpCode" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(236): error: identifier "s_boolean_binary_ew_op_codes" is undefined

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

/home/wnitsch/src/Open3D/src/Open3D/Core/Kernel/BinaryEWCUDA.cu(257): error: name followed by "::" must be a class or namespace name

69 errors detected in the compilation of "/tmp/tmpxft_00007cdb_00000000-4_BinaryEWCUDA.cpp4.ii".
src/Open3D/Core/CMakeFiles/Core.dir/build.make:335: recipe for target 'src/Open3D/Core/CMakeFiles/Core.dir/Kernel/BinaryEWCUDA.cu.o' failed
make[2]: *** [src/Open3D/Core/CMakeFiles/Core.dir/Kernel/BinaryEWCUDA.cu.o] Error 2
CMakeFiles/Makefile2:1298: recipe for target 'src/Open3D/Core/CMakeFiles/Core.dir/all' failed
make[1]: *** [src/Open3D/Core/CMakeFiles/Core.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2

```

--- a/src/Open3D/Core/Indexer.cpp
+++ b/src/Open3D/Core/Indexer.cpp
@@ -41,11 +41,24 @@ Indexer::Indexer(const std::vector<Tensor>& input_tensors,
                  const std::vector<Tensor>& output_tensors,
                  DtypePolicy dtype_policy,
                  const SizeVector& reduction_dims) {
-    if (input_tensors.size() < 1) {
+    // Check the number of inputs and outputs.
+    num_inputs_ = static_cast<int64_t>(input_tensors.size());
+    num_outputs_ = static_cast<int64_t>(output_tensors.size());
+    if (num_inputs_ < 1) {
         utility::LogError("Indexer must have at least one input.");
     }
-    if (output_tensors.size() < 1) {
+    if (num_inputs_ > MAX_INPUTS) {
+        utility::LogError(
+                "Indexer cannot have more than {} inputs, but got {}.",
+                MAX_INPUTS, num_inputs_);
+    }
+    if (num_outputs_ < 1) {
         utility::LogError("Indexer must have at least one input.");
+    }
+    if (num_outputs_ > MAX_OUTPUTS) {
+        utility::LogError(
+                "Indexer cannot have more than {} outputs, but got {}.",
+                MAX_OUTPUTS, num_outputs_);
     }
 
     // Dtype sanity check and handling.
@@ -100,6 +113,14 @@ Indexer::Indexer(const std::vector<Tensor>& input_tensors,
         utility::LogError("Unimplemented dtype policy");
     }
 
+    // Convert to TensorRef.
+    for (int64_t i = 0; i < num_inputs_; ++i) {
+        inputs_[i] = TensorRef(input_tensors[i]);
+    }
+    for (int64_t i = 0; i < num_outputs_; ++i) {
+        outputs_[i] = TensorRef(output_tensors[i]);
+    }
+
     // For simplicity, all outputs must have the same shape.
     SizeVector ref_output_shape = output_tensors[0].GetShape();
     for (const auto& output_tensor : output_tensors) {
@@ -109,25 +130,6 @@ Indexer::Indexer(const std::vector<Tensor>& input_tensors,
                     "but {} != {}",
                     output_tensor.GetShape(), ref_output_shape);
         }
-    }
-
-    // Convert to TensorRef.
-    num_inputs_ = static_cast<int64_t>(input_tensors.size());
-    if (num_inputs_ > MAX_OPERANDS) {
-        utility::LogError("Operation has too many inputs {} > {}", num_inputs_,
-                          MAX_OPERANDS);
-    }
-    for (int64_t i = 0; i < num_inputs_; ++i) {
-        inputs_[i] = TensorRef(input_tensors[i]);
-    }
-
-    num_outputs_ = static_cast<int64_t>(output_tensors.size());
-    if (num_outputs_ > MAX_OPERANDS) {
-        utility::LogError("Operation has too many outputs {} > {}",
-                          num_outputs_, MAX_OPERANDS);
-    }
-    for (int64_t i = 0; i < num_outputs_; ++i) {
-        outputs_[i] = TensorRef(output_tensors[i]);
     }
 
     // Theoretically, reduction can be mixed with broadcasting. For

--- a/src/Open3D/Core/Indexer.h
+++ b/src/Open3D/Core/Indexer.h
@@ -44,8 +44,13 @@ class IndexerIterator;
 // Maximum number of dimensions of TensorRef.
 static constexpr int64_t MAX_DIMS = 10;
 
-// Maximum number of operands (inputs) of an op.
-static constexpr int64_t MAX_OPERANDS = 10;
+// Maximum number of inputs of an op.
+// MAX_INPUTS shall be >= MAX_DIMS to support advanced indexing.
+static constexpr int64_t MAX_INPUTS = 10;
+
+// Maximum number of outputs of an op. This number can be increased when
+// necessary.
+static constexpr int64_t MAX_OUTPUTS = 2;
 
 // Fixed-size array type usable from host and device.
 template <typename T, int size>
@@ -446,10 +451,10 @@ protected:
     int64_t num_outputs_ = 0;
 
     /// Array of input TensorRefs.
-    TensorRef inputs_[MAX_OPERANDS];
+    TensorRef inputs_[MAX_INPUTS];
 
     /// Output TensorRef.
-    TensorRef outputs_[MAX_OPERANDS];
+    TensorRef outputs_[MAX_OUTPUTS];
 
     /// Indexer's global shape. The shape's number of elements is the
     /// same as GetNumWorkloads() for the Indexer.

--- a/src/Open3D/Core/Tensor.h
+++ b/src/Open3D/Core/Tensor.h
@@ -145,6 +145,11 @@ public:
         return *this;
     }
 
+    /// \brief Fill the whole Tensor with a scalar value, the scalar will be
+    /// casted to the Tensor's dtype.
+    template <typename T>
+    void Fill(T v);
+
     /// Create a tensor with uninitilized values.
     static Tensor Empty(const SizeVector& shape,
                         Dtype dtype,
@@ -258,18 +263,6 @@ public:
     /// change. Slices of the original Tensor still keeps the original memory.
     /// After assignment, the Tensor will be contiguous.
     void Assign(const Tensor& other);
-
-    /// \brief Fill the whole Tensor with a scalar value, the scalar will be
-    /// casted to the Tensor's dtype.
-    template <typename T>
-    void Fill(T v) {
-        DISPATCH_DTYPE_TO_TEMPLATE_WITH_BOOL(GetDtype(), [&]() {
-            scalar_t casted_v = static_cast<scalar_t>(v);
-            Tensor tmp(std::vector<scalar_t>({casted_v}), SizeVector({}),
-                       GetDtype(), GetDevice());
-            AsRvalue() = tmp;
-        });
-    }
 
     /// Broadcast Tensor to a new broadcastable shape.
     Tensor Broadcast(const SizeVector& dst_shape) const;
@@ -854,23 +847,33 @@ inline std::vector<bool> Tensor::ToFlatVector() const {
     return values;
 }
 
+template <typename Scalar>
+inline void Tensor::Fill(Scalar v) {
+    DISPATCH_DTYPE_TO_TEMPLATE_WITH_BOOL(GetDtype(), [&]() {
+        scalar_t casted_v = static_cast<scalar_t>(v);
+        Tensor tmp(std::vector<scalar_t>({casted_v}), SizeVector({}),
+                   GetDtype(), GetDevice());
+        AsRvalue() = tmp;
+    });
+}
+
 template <typename T>
-Tensor operator+(T scalar_lhs, const Tensor& rhs) {
+inline Tensor operator+(T scalar_lhs, const Tensor& rhs) {
     return rhs + scalar_lhs;
 }
 
 template <typename T>
-Tensor operator-(T scalar_lhs, const Tensor& rhs) {
+inline Tensor operator-(T scalar_lhs, const Tensor& rhs) {
     return Tensor::Full({}, scalar_lhs, rhs.GetDtype(), rhs.GetDevice()) - rhs;
 }
 
 template <typename T>
-Tensor operator*(T scalar_lhs, const Tensor& rhs) {
+inline Tensor operator*(T scalar_lhs, const Tensor& rhs) {
     return rhs * scalar_lhs;
 }
 
 template <typename T>
-Tensor operator/(T scalar_lhs, const Tensor& rhs) {
+inline Tensor operator/(T scalar_lhs, const Tensor& rhs) {
     return Tensor::Full({}, scalar_lhs, rhs.GetDtype(), rhs.GetDevice()) / rhs;
 }
 

